### PR TITLE
Substitute GT_RET_EXPR in inline candidate arguments

### PR DESCRIPTION
**Pull request by jakobbotsch**
These `GT_RET_EXPR` nodes would previously make it into the inlinees as part of the inline argument table. This was confusing to me, and it additionally means that we call `gtRetExprVal()` in a few places to compensate (while for non-toplevel cases it would essentially cause the argument node containing the `GT_RET_EXPR` to become "opaque" to the JIT). With this change we only ever expect to substitute `GT_RET_EXPR` inside `fgUpdateInlineReturnExpressionPlaceHolder` called from the main `fgInline` loop, so I have inlined `gtRetExprVal()` here and removed that function.

### CODE REVIEW

File: src/coreclr/jit/fginline.cpp [hunk 1]
```diff
@@ -138,6 +138,16 @@ PhaseStatus Compiler::fgInline()
                 // they belong.
                 if (call->IsInlineCandidate() || call->IsGuardedDevirtualizationCandidate())
                 {
+                    // Yes. We may still have GT_RET_EXPR in arguments,
+                    // substitute those here so the inlinee does not need to
+                    // deal with them.
+                    // Note that we leave late devirts in arguments for when we
+                    // handle inlined statements as fgLateDevirtualization
+                    // cannot handle being called twice on a late devirt
+                    // candidate.
```
[the remark this answers is not in the document]
Comment (jakobbotsch) [reply src/coreclr/jit/fginline.cpp]: Decided to fix this, didn't turn out to be too difficult.

**Comment by jakobbotsch** [comment]
Fuzzlyn failures are preexisting. libraries-jitstress failures are #69154 and #68803.

[Diffs.](https://dev.azure.com/dnceng/public/_build/results?buildId=1762370&view=ms.vss-build-web.run-extensions-tab) Generally we see small improvements in the non-test code, and some regressions in the test code. Presumably these are because we now have fewer "opaque" nodes during inlining. I will spot check these when I have some more time, but I don't think we need to block the review on them.

There are some TP regressions, but they are mostly paid for by #69120 that I noticed while working on this. I think there are additional improvements to be made in the future; in particular, we currently walk the argument nodes once when we see the inline candidate, and then another time after they are added as part of inlining (because `fgInlinePrependStatements` adds them after the inline candidate call). So I think we can afford this given that it makes follow-up improvements simpler and generally simplifies where `GT_RET_EXPR` appears.

cc @dotnet/jit-contrib PTAL @AndyAyersMS

**Comment by AndyAyersMS** [comment]
I am curious about what's behind some of the larger diffs here, so please do take a look
```
         172 (15.45 % of base) : 21274.dasm - RequestUtilities:AddHeader(HttpRequestMessage,String,StringValues)
```

**Comment by jakobbotsch** [comment]
> I am curious about what's behind some of the larger diffs here, so please do take a look

Hmm yes. And after looking at the diff summary again I can see that

> Generally we see small improvements in the non-test code, and some regressions in the test code.

is no longer true. That's interesting because it was true before my latest commit 41e0af21086d885f37058e6430828f46279545f1, that changes late devirt to happen on the call arguments as well. [These](https://dev.azure.com/dnceng/public/_build/results?buildId=1762088&view=ms.vss-build-web.run-extensions-tab) were the diffs before that.

File: src/coreclr/jit/importer.cpp [hunk 2]
```diff
@@ -21299,6 +21300,7 @@ void Compiler::impDevirtualizeCall(GenTreeCall*            call,
     // it's a union field used for other things by virtual
     // stubs)
     call->gtInlineCandidateInfo = nullptr;
+    call->gtCallMoreFlags &= ~GTF_CALL_M_LATE_DEVIRT;
```
Comment (AndyAyersMS) [inline src/coreclr/jit/importer.cpp:21303]: Maybe we should rename this flag? `GTF_CALL_M_HAS_LATE_DEVIRT_INFO` perhaps? I introduced it but the name could better describe what it means.

**Review by AndyAyersMS — COMMENTED** [review]
Changes look good.

I wonder if the diffs here are because this change uncovers more devirt + inlines?

**Comment by jakobbotsch** [comment]
> I am curious about what's behind some of the larger diffs here, so please do take a look
> 
> ```
>          172 (15.45 % of base) : 21274.dasm - RequestUtilities:AddHeader(HttpRequestMessage,String,StringValues)
> ```

The difference here is that we do another inline because of:
```diff
-Inline candidate callsite is boring.  Multiplier increased to 2.3.
+Inline candidate callsite is warm.  Multiplier increased to 3.
```
This happens because we now do the following substitution:
```diff
+Replacing the return expression placeholder [000164] with [000812]
+               [000164] --C--------                         ▌  RET_EXPR  ref   (inl return expr [000812])
+
+Inserting the inline return expression
+               [000812] ---XG------                         ▌  FIELD     ref    _headers
+               [000811] -----------                         └──▌  LCL_VAR   ref    V13 tmp6         
+
 Expanding INLINE_CANDIDATE in statement STMT00050 in BB105:
 STMT00050 ( ??? ... ??? )
                [000167] I-C-G------                         ▌  CALL nullcheck int    HttpHeaders.TryAddWithoutValidation (exactContextHnd=0x00007FFC32571731)
-                             [000164] --C-------- this                    ├──▌  RET_EXPR  ref   (inl return expr [000812])
+               [000812] ---XG------ this                    ├──▌  FIELD     ref    _headers
+               [000811] -----------                         │  └──▌  LCL_VAR   ref    V13 tmp6         
                [000165] ----------- arg1                    ├──▌  LCL_VAR   ref    V01 arg1         
                [000166] ----------- arg2                    └──▌  LCL_VAR   ref    V03 loc0         
```

And that `RET_EXPR` node has `BBF_PROF_WEIGHT` which we propagate in `fgUpdateInlineReturnExpressionPlaceHolder`.
If I understand correctly, the same thing would have happened if the `RET_EXPR` was in the statement before the inline candidate, so this doesn't seem bad to me.
I'll spot check some more diffs.

**Comment by jakobbotsch** [comment]
The above may also explain parts of the TP diff, since if we do more inlines we expect to spend more time.

**Comment by jakobbotsch** [comment]
In asm.benchmarks.run.windows.x64.checked 8113 we have
```diff
-; 0 inlinees with PGO data; 59 single block inlinees; 39 inlinees without PGO data
+; 0 inlinees with PGO data; 69 single block inlinees; 48 inlinees without PGO data
```

The cause is that we create a couple fewer locals, which means in the new version we do one less lva table grow (which has a *= 2 multiplier). Then we end up with a bunch of extra inlines due to space in the lva table:
```diff
-Caller has 98 locals.  Multiplier decreased to 10.2186.
+Caller has 148 locals.  Multiplier decreased to 9.6668.
 calleeNativeSizeEstimate=1122
 callsiteNativeSizeEstimate=115
-benefit multiplier=9.6668
+benefit multiplier=10.2186
-threshold=1111
+threshold=1175
-Native estimate for function size exceeds threshold for inlining 112.2 > 111.1 (multiplier = 9.6668)
+Native estimate for function size is within threshold for inlining 112.2 <= 117.5 (multiplier = 10.2186)
```

Seems strange that we are using the capacity here:
https://github.com/dotnet/runtime/blob/96db556687a08b65982abbfc76256e7312ffe994/src/coreclr/jit/inlinepolicy.cpp#L1708-L1714
Should this be using `lvaCount` instead?

**Comment by AndyAyersMS** [comment]
> Should this be using `lvaCount` instead?

I think this is something @egorbo added... yeah, it seems odd to check capacity and not how much is actually in use.

File: src/coreclr/jit/importer.cpp (hunk 2 above)
Comment (jakobbotsch) [reply src/coreclr/jit/importer.cpp]: Done.

**Comment by jakobbotsch** [comment]
Timeouts are known. Can you please take a look at the above diffs and approve if it sounds fine @AndyAyersMS?

**Comment by jakobbotsch** [comment]
Ping @AndyAyersMS

**Comment by AndyAyersMS** [comment]
Sorry, got caught up in other things. Looking now.

**Review by AndyAyersMS — APPROVED** [review]
LGTM. Thanks for looking at diffs.

**Pull request merged**